code/M3D/m3d.h

```diff
@@ -0,0 +1,4544 @@
+
+/* callbacks */
+typedef unsigned char *(*m3dread_t)(char *filename, unsigned int *size);                        /* read file contents into buffer */
+typedef void (*m3dfree_t)(void *buffer);                                                        /* free file contents buffer */
+typedef int (*m3dtxsc_t)(const char *name, const void *script, uint32_t len, m3dtx_t *output);  /* interpret texture script */
+typedef int (*m3dprsc_t)(const char *name, const void *script, uint32_t len, m3d_t *model);     /* interpret surface script */
+#endif /* ifndef M3D_APIVERSION */
+
+/*** C prototypes ***/
+/* import / export */
+m3d_t *m3d_load(unsigned char *data, m3dread_t readfilecb, m3dfree_t freecb, m3d_t *mtllib);
+unsigned char *m3d_save(m3d_t *model, int quality, int flags, unsigned int *size);
+void m3d_free(m3d_t *model);
+/* generate animation pose skeleton */
+m3dtr_t *m3d_frame(m3d_t *model, M3D_INDEX actionid, M3D_INDEX frameid, m3dtr_t *skeleton);
+m3db_t *m3d_pose(m3d_t *model, M3D_INDEX actionid, uint32_t msec);
+
+/* private prototypes used by both importer and exporter */
+m3ds_t *_m3d_addskin(m3ds_t *skin, uint32_t *numskin, m3ds_t *s, uint32_t *idx);
+m3dv_t *_m3d_addnorm(m3dv_t *vrtx, uint32_t *numvrtx, m3dv_t *v, uint32_t *idx);
+char *_m3d_safestr(char *in, int morelines);
+
+/*** C implementation ***/
+#ifdef M3D_IMPLEMENTATION
+#if !defined(M3D_NOIMPORTER) || defined(M3D_EXPORTER)
+/* material property definitions */
+static m3dpd_t m3d_propertytypes[] = {
+    M3D_PROPERTYDEF(m3dpf_color, m3dp_Kd, "Kd"),    /* diffuse color */
+    M3D_PROPERTYDEF(m3dpf_color, m3dp_Ka, "Ka"),    /* ambient color */
+    M3D_PROPERTYDEF(m3dpf_color, m3dp_Ks, "Ks"),    /* specular color */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Ns, "Ns"),    /* specular exponent */
+    M3D_PROPERTYDEF(m3dpf_color, m3dp_Ke, "Ke"),    /* emissive (emitting light of this color) */
+    M3D_PROPERTYDEF(m3dpf_color, m3dp_Tf, "Tf"),    /* transmission color */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Km, "Km"),    /* bump strength */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_d,  "d"),     /* dissolve (transparency) */
+    M3D_PROPERTYDEF(m3dpf_uint8, m3dp_il, "il"),    /* illumination model (informational, ignored by PBR-shaders) */
+
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Pr, "Pr"),    /* roughness */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Pm, "Pm"),    /* metallic, also reflection */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Ps, "Ps"),    /* sheen */
+    M3D_PROPERTYDEF(m3dpf_float, m3dp_Ni, "Ni"),    /* index of refraction (optical density) */
+
+    /* aliases, note that "map_*" aliases are handled automatically */
+    M3D_PROPERTYDEF(m3dpf_map, m3dp_map_Km, "bump"),
+    M3D_PROPERTYDEF(m3dpf_map, m3dp_map_Pm, "refl")
+};
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#if !defined(M3D_NOIMPORTER) && !defined(STBI_INCLUDE_STB_IMAGE_H)
+/* PNG decompressor from
+
+   stb_image - v2.23 - public domain image loader - http://nothings.org/stb_image.h
```

Comment:
Did you really need to copy-paste the header? Why not include it as a separate file? At least it would make it easier to update. Also, do we export these symbols from the dynamic object? What happens if user code also uses stb_image?

Reply:
As the comment says, it is not the entire header, just part of it (which by the way is also needed by this format, not just for PNG). And about including, the purpose of dependency-free single file headers is exactly that, "not to include as a separate file". Don't be afraid, the ifdef guards are there, so if it is included in a separate file it will work as expected.

Reply:
@turol Just ran into your exact hypothetical scenario. My code uses `stb_image` and then conflicts with these symbols here. Why does it matter if we include the entire header or not? You're only supposed to define `STB_IMAGE_IMPLEMENTATION` once. #2766

Reply:
Because at that point two symbols (functions) exist with the same name. The user code defines one set (with `STB_IMAGE_IMPLEMENTATION`, used exactly once as you're supposed to) and your copy-pasted code defines the other. If the dynamic library doesn't export these symbols (on windows it shouldn't, on Linux and OS X it's possible to do it but I don't know if that's been done here) then no problem. If does or if using static library then duplicate symbols error happens.

Two ways to fix this:
1. Rename your copy-pasted symbols so none of them clash. Leads to duplicate code, makes it hard to update it and can bloat the executable.
2. Replace your code with the real `stb_image.h`, include one object file where you define `STB_IMAGE_IMPLEMENTATION `. Document that it's in there so users aren't caught unaware. Provide CMake option to disable it in case users want to use theirs.

Reply:
@turol Correct, we are defining it once in our user code and linking Assimp statically through Hunter, resulting in the duplicate symbol issue. To be clear, we are just consuming the latest version of assimp and are not contributing the M3D code. It looks like #2766 is implementing solution (1), which I agree is adding bloat and less maintainable. (2) would be preferred. Thoughts @bztsrc?